text/0000-externally-implementable-functions.md

```diff
@@ -0,0 +1,151 @@
+
+```rust
+// In another crate:
+
+impl fn x::a() {
+    println!("my implementation of a");
+}
+
+impl fn x::b() {
+    println!("my implementation of b");
+}
+```
+
+# Details
+
+## Signature
+
+It is an error to have a different signature for the `impl fn` item.
+
+(Whether `#[track_caller]` is used or not is considered part of the signature here.)
+
+## No impl
+
+It is an error to have no `impl fn` item (in any crate) for an `extern impl fn` item without a body.
+
+## Duplicates
+
+It is an error to have multiple `impl fn` items (across all crates) for the same `extern impl fn` item.
+
+## Visibility
+
+`extern impl fn` items can have a visibility specifier (like `pub`), which determines who can *call* the function (or create pointers to it, etc.).
+
+*Implementing* the function can be done by any crate that can name the item.
+(The `impl fn` item will need to name the item to implement, which could be directly or through an alias/re-export.)
+
+# Implementation
+
+The implementation will be based on the same mechanisms as used today for the `panic_handler` and `#[global_allocator]` features.
+
+The compiler of the root crate will find the implementation of all externally implementable functions and give an error
+if more than one implementation is found for any of them.
+If none are found, the result is either an error, or—if the `extern impl fn` has a default body—an implementation
+is generated that calls that default body.
+
+# Drawbacks
+
+- It encourages globally defined behaviour.
+  - Counterargument: We are already doing this anyway, both inside the standard library (e.g. panic_handler, allocator)
+    and outside (e.g. global logger). This just makes it much easier (and safer) to get right.
+- This will invite the addition of many hooks to the standard library to modify existing behavior.
+  While we should consider such possibilities, this RFC does not propose that every piece of standard library behavior should be replaceable.
+
+# Rationale and alternatives
+
+- The syntax re-uses existing keywords. Alternatively, we could:
+  - Use the `override` reserved keyword.
+  - Add a new (contextual) keyword (e.g. `existential fn`).
+  - Use an attribute (e.g. `#[extern_impl]`) instead.
+
+# Prior art
+
+[RFC 2494 "Existential types with external definition"](https://github.com/rust-lang/rfcs/pull/2492)
+has been proposed before, which basically does this for *types*. Doing this for functions (as a start) saves a lot of complexity.
+
+# Unresolved questions
+
+- Should we provide a mechanism to set an `extern impl fn` using `=` from an existing `fn` value, rather than writing a body? For instance, `impl fn x::y = a::b;`
+- Should we allow some form of subtyping, similarly to how traits allow trait impls to do subtyping?
+- What should the syntax be once we stabilize this?
+- How should this work in dynamic libraries?
+- Should not having an implementation be an error when the function is never called (after dead code elimination)?
```

Comment:
Leaving aside the issues with optimization-dependence, there is a *lot* of value in being able to figure out cases like the following:

- Crate `a` provides `extern impl fn hook()`, with no default.
- Crate `a` provides various APIs, only one of which, `uses_hook()`, ever calls `hook()`.
- Binary crate `b` depends on `a`.
- `b` does not provide an impl of `hook()`.
- `b` never calls `uses_hook()`.

I think this case should link and work. I see many use cases for "if you want to call X you need to provide an impl of Y, but you don't have to provide an impl of Y if you never call X".

To be clear, the case I'm concerned about does not depend on optimization-based dead-code elimination (e.g. `if false { uses_hook(); }`). The case I'm concerned about is where it's *statically* obvious to the compiler that nothing calls `uses_hook()` (either because nothing calls it directly, or because anything calling it directly is itself never called).

Reply:
I fear that by the point at which the knowledge (that the hook is never used) becomes available, we could already be past some early error stages, making this a very late error to discover. In any case, crate a could always feature-gate the hook and all relevant methods to make such a case explicit.

Reply:
Supporting this would require delaying codegen of `uses_hook()` to all call sites. Otherwise without `--gc-sections` (because of `-Clink-dead-code=yes` or because of a linker which doesn't support `--gc-sections`) the linker will complain about the missing symbol definition.

It also leaks the internal implementation details of functions unless the function signature is explicitly required to mention all externally implementable functions that are transitively used.

Reply:
@joshtriplett How do you feel about the fact that one needs to provide a `#[panic_handler]` in a no_std program even if the entire crate is just `pub fn start() {}` or something like that, where nothing can possibly call the panic handler?

I agree there could be value in only requiring it when necessary, but I think there is also a lot of value in keeping this feature consistent with how things like `#[panic_handler]` work.

Reply:
@m-ou-se I can appreciate that it's hard for the compiler to statically determine "nothing ever panics and thus nothing ever uses the panic handler". The panic handler in particular seems potentially difficult in that way.

It *seems* like it should be easier to determine "nothing ever calls this function".

@bjorn3
> Supporting this would require delaying codegen of `uses_hook()` to all call sites. Otherwise without `--gc-sections` (because of `-Clink-dead-code=yes` or because of a linker which doesn't support `--gc-sections`) the linker will complain about the missing symbol definition.

I think it'd be *perfectly fine* if this errored out if you use `-Clink-dead-code=yes`, on the basis that the dead code you're including has calls to the externally implementable function. Does that make it easier?

-----

That said, I can see another possible way to handle this, as well. If a crate wants to make a set of functionality only available given a specific externally implementable function, it can put *both* the functionality and the externally implementable function behind the same feature flag, and tell users to turn that feature on and then provide the function. That also addresses @bjorn3's point about exposing implementation details, since the feature flag would define the set of things that need the externally implementable function.

Given that, I'll submit a suggestion documenting this, and move this case to future work.